pkg/cli/nodes_connectivity.go

```diff
@@ -0,0 +1,403 @@
+package cli
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/kustomize/api/types"
+
+	"github.com/google/uuid"
+	"github.com/replicatedhq/plumber/v2"
+	"github.com/spf13/cobra"
+
+	"github.com/replicatedhq/kurl/pkg/k8sutil"
+	"github.com/replicatedhq/kurl/pkg/static/nodes_connectivity"
+)
+
+var usageExamples = `
+# Test if all nodes can reach all other nodes using tcp in port 6472.
+kurl netutil nodes-connectivity --port 6472 --proto tcp
+# Test if all nodes can reach all other nodes using udp in port 7788.
+kurl netutil nodes-connectivity --port 7788 --proto udp
+`
+
+const (
+	listenersSelector = "name=nodes-connectivity-listener"
+	pingerSelector    = "name=nodes-connectivity-pinger"
+)
+
+type logLine struct {
+	message string
+	err     error
+}
+
+type nodeConnectivityOptions struct {
+	namespace string
+	proto     string
+	attempts  int
+	cliset    kubernetes.Interface
+	image     string
+	port      int32
+	cli       client.Client
+	wait      time.Duration
+}
+
+func newNetutilNodesConnectivity(_ CLI) *cobra.Command {
+	var opts nodeConnectivityOptions
+	cmd := &cobra.Command{
+		Use:     "nodes-connectivity",
+		Short:   "Tests if all nodes can reach all other nodes using the provided protocol in the provided port",
+		Example: usageExamples,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), syscall.SIGTERM, syscall.SIGINT)
+			defer cancel()
+
+			if opts.port == 0 {
+				return fmt.Errorf("--port flag is required.")
+			}
+			opts.proto = strings.ToUpper(opts.proto)
+			proto := corev1.Protocol(opts.proto)
+			if proto != corev1.ProtocolTCP && proto != corev1.ProtocolUDP {
+				return fmt.Errorf("--protocol must be either tcp or udp.")
+			}
+			opts.wait = time.Second
+			if corev1.Protocol(opts.proto) == corev1.ProtocolUDP {
+				opts.wait = 5 * time.Second
+			}
+
+			cli, err := client.New(config.GetConfigOrDie(), client.Options{})
+			if err != nil {
+				return fmt.Errorf("failed to create kubernetes client: %w", err)
+			}
+			opts.cli = cli
+
+			// XXX controller-runtime kubernetes client does not have access to subresources that span beneath
+			// crud objects (e.g. pod logs). we have to create a kubernetes clientset, and that is why we can't
+			// have nice things.
+			cliset, err := kubernetes.NewForConfig(config.GetConfigOrDie())
+			if err != nil {
+				log.Printf("Failed to create kubernetes client set: %s", err)
+				os.Exit(1)
```

Comment:
See my previous [comment](https://github.com/replicatedhq/kURL/pull/4376#discussion_r1166308321). Essentially in some cases I thought it would be useful to print the "Usage" while not in others. I am still adjusting this.